Assorted spec and implementation work

xeno-by opened the pull request:
Summary:
  * Implemented the new method symbol format (#1295)
  * Specified and implemented `SELF_PARAMETER`.
  * Codified pattern variables.
  * Codified type variables.

Todo: 
  * [X] Implement the new method symbol format.
  * [ ] Implement the new encoding for vals and vars.
  * [ ] Decide whether `hasNontrivialPrefix` should stay or go.

olafurpg commented on tests/jvm/src/test/resources/lowlevel.expect:
```diff
@@ -345,9 +359,344 @@ Occurrences:
+  Methods => _root_.example.Methods#
+  T => _root_.example.Methods#[T]
+  AList => _root_.example.Methods#AList#
+_root_.example.Methods#m11a(.type). => def m11a: (x: Predef.type): Nothing
```
Missing test for disambiguating singleton types

```
class Foo[T]  {
  val x = ""
  object n { val x = List(1) }
  def m6(a: x.type) = ???
  def m6(b: n.x.type) = ???
}
```

and also test at reference site.

xeno-by replied:
Done.

olafurpg commented on semanticdb/integration/src/main/scala/example/Methods.scala:
```diff
@@ -0,0 +1,30 @@
+  def m6(x: Int) = ???
+  def m6(x: List[T]) = ???
+  def m6(x: scala.List[T]) = ???
+  def m7[U: Ordering](c: Methods[T], l: List[U]) = ???
```
Missing test for context view

```scala
def m7[U <% Ordering[T]] = ???
def m8[U <% Foo] = ???
```

xeno-by replied:
Done.

olafurpg commented on semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/Main.scala:
```diff
@@ -447,28 +404,58 @@ object Main {
+    def disambiguator: String = {
+      val kindred = sym.parent.get.children.filter(other => skind(other) == skind(sym))
+      val siblings = kindred.filter(_.name == sym.name)
+      val synonyms = siblings.filter(_.descriptor == sym.descriptor)
```
Do we have any ordering guarantees on `Symbol.children`? What is the chance that `+1` and `+2` get mixed up?

xeno-by replied:
To the best of my knowledge, pickling preserves the order of symbols in scopes. Scalac namer enters symbols in the order of their appearance in source code. Also see https://github.com/retronym/scalac-stability/issues/1.

MasseGuillaume commented:
It would be nice to show how you can retrieve a conflicting symbol (for example `m6`) in the TargetedSuite.

olafurpg commented on tests/jvm/src/test/scala/scala/meta/tests/semanticdb/TargetedSuite.scala:
```diff
@@ -93,8 +93,8 @@ class TargetedSuite extends DatabaseSuite(SemanticdbMode.Slim) {
     """.trim.stripMargin, { (db, copy, age) =>
-      assert(copy === Symbol("_root_.e.User#copy(Ljava/lang/String;I)Le/User;."))
-      assert(age === Symbol("_root_.e.User#copy(Ljava/lang/String;I)Le/User;.(age)"))
+      assert(copy === Symbol("_root_.e.User#copy(String,Int)."))
```
Can we add a targeted test like this here to highlight the disambiguation like `List+1`/`List+2`? It's covered in the expect tests but I think it deserves highlighting in the targeted suite as @MasseGuillaume pointed out.

xeno-by replied:
Done.

xeno-by commented on semanticdb/integration/src/main/scala/example/Methods.scala:
```diff
@@ -0,0 +1,30 @@
+  def m5(x: Int) = ???
+  def m6(x: Int) = ???
+  def m6(x: List[T]) = ???
+  def m6(x: scala.List[T]) = ???
```
Done.

xeno-by commented on semanticdb/integration/src/main/scala/example/Methods.scala:
```diff
@@ -0,0 +1,30 @@
+import scala.math.Ordering
+
+class Methods[T] {
+  class List[T]
```
Noted. If you don't mind, I'd like to keep `List` as is in these tests, since changing it would be a big diff. In the future, I'll stay away from it.

xeno-by commented:
@MasseGuillaume @olafurpg Addressed your feedback, except for references to java symbols.

Our specification does not support Java at the moment, so the implementation is free to do whatever. Support for Java is on our board though - https://github.com/scalameta/scalameta/issues/1288 - so I expect we'll get there soon.
